sandy/fy.py

```diff
@@ -357,6 +360,147 @@ def apply_qmatrix(self, zam, energy, decay_data):
+            V_x = CFY.pivot_table(index='ZAP', columns='ZAP', values='DFY',
+                                  aggfunc='sum').fillna(0)
+            V_y = V_y.pivot_table(index='ZAP', columns='ZAP', values='DFY',
+                                  aggfunc='sum').fillna(0)
+        # Calculate the sensitivity:
+
+        def _GLS_sensitivity(S, V_y, V_x, threshold=None):
+            M = S.dot(V_x).dot(S.T) + V_y
+            M_calc = sandy.core.cov._Cov(M)
+            M_nonzero_idxs, M_reduce = M_calc._reduce_size()
+            unit = np.identity(len(M_reduce))
+            M_reduce_inverse = splu(csc_matrix(M_reduce)).solve(unit)
+            M_inv = pd.DataFrame(
+                sandy.core.cov._Cov._restore_size(M_nonzero_idxs, M_reduce_inverse, len(M_calc)),
+                index=S.index,
+                columns=S.index
+                                 )
+            sensitivity = V_x.dot(S.T).dot(M_inv)
+            if threshold is not None:
+                sensitivity[sensitivity < threshold] = 0
+            return sensitivity
+        return _GLS_sensitivity(S, V_y, V_x, threshold=1.0e-14)
+
+    def custom_GLS(self, zam, e, perturbed_fy, sensitivity, pert_data):
+        """
+        Apply a GLS method perturbation in the fission yields original data
+        for a given zam in a given energy.
+
+        Parameters
+        ----------
+        zam : `int`
+            ZAM number of the material to which calculations are to be
+            applied.
+        e : `float`
+            Energy to which calculations are to be applied.
+        perturbed_fy : `Fy` object
+            Fy object with the perturbed CFY or IFY data.
+        sensitivity : `pd.DataFrame`
+            GLS sensitivity matrix for that MT perturbed.
+        pert_data : `int`
+            The MT number of the perturbed data.
+
+        Raises
+        ------
+        sandy.error
+            Perturbed data MT is not in Fy object.
+
+        Returns
+        -------
+        `Fy`
+            Fission yield instance with IFY and CFY perturbed using GLS
+            for a given ZAM/e/decay_data.
+
+        """
+        new_data = self.data.copy()
+        mask = (new_data.ZAM == zam) & (new_data.E == e)
+        # Divide the CFY and IFY:
+        fy_type = perturbed_fy.data.loc[mask].groupby('MT')
+        IFY_pert, CFY_pert = [fy_type.get_group(x)[['ZAP', 'FY']]
+                              .set_index('ZAP') for x in fy_type.groups]
+        fy_type = new_data.loc[mask].groupby('MT')
+        IFY, CFY = [fy_type.get_group(x)[['ZAP', 'FY']]
+                    .set_index('ZAP') for x in fy_type.groups]
+        # perform the GLS method
+        if pert_data == 454:
+            CFY_new = CFY + sensitivity.dot(IFY_pert-IFY)
+            CFY_new[CFY_new < 1.0e-14] = 0  # Avoid numerical fluctuations
+            CFY = CFY_new
+            IFY = IFY_pert
+        elif pert_data == 459:
+            IFY_new = IFY + sensitivity.dot(CFY_pert-CFY)
```

Comment:
Now I understand better what you do...

I think a fundamental issue is that you use variable name `CFY_pert`, while this does not need to be a *perturbed* parameter